std/container/binaryheap.d

```diff
@@ -369,12 +369,40 @@ must be collected.
             insert(value);
             return true;
         }
-        // must replace the top
+
+        assert(!_store.empty, "Cannot replace front of an empty heap.");
+        if (!comp(value, _store.front)) return false; // value >= largest
+        _store.front = value;
+
+        percolate(_store[], 0, _length);
+        debug(BinaryHeap) assertValid();
+        return true;
+    }
+
+/**
+If the heap has room to grow, inserts $(D value) into the store and
+returns $(D true). Otherwise, if $(D less(value, front)), copies front
+in value and returns $(D true). Otherwise, leaves
+the heap unaffected and returns $(D false).
+     */
+    bool conditionalSwap(ref ElementType!Store value)
+    {
+        _payload.refCountedStore.ensureInitialized();
+        if (_length < _store.length)
+        {
+            insert(value);
```

Comment:
I have no idea what the behavior should be if the heap has room to grow. The problem here is we collapse together the case when the heap is full and we swap an element with another, with the case when the heap is still growing.

Let's do this for now: `assert(_length == _store.length)` in the code, and mention that it is a precondition that the heap is fully occupied. The advantage of this is that we later get to define this behavior if needed; whereas if we define it the wrong way now, we cannot change it later.